File: src/Meshboard.Web/lib/models/external-issue.ts
export type ExternalIssueActorModel = {
    name: string;
    username?: string | null;
    externalUserId?: string | null;
    externalUsername?: string | null;
    externalDisplayName?: string | null;
    meshboardUserId?: string | null;
    meshboardUsername?: string | null;
    meshboardDisplayName?: string | null;
};

export type BoardIssueSummaryModel = {
    externalId: string;
    detailsLookupKey: string;
    issueNumber: string;
    sourceKey: string;
    sourceName?: string | null;
    title: string;
    descriptionPreview?: string | null;
    status: string;
    url?: string | null;
    assignee?: ExternalIssueActorModel | null;
    sourceColumn?: string | null;
    boardColumnId?: string | null;
    createdAt?: string | null;
    updatedAt?: string | null;
    labels: string[];
};

export type ExternalIssueModel = {
    externalId: string;
    detailsLookupKey: string;
    issueNumber: string;
    sourceKey: string;
    sourceName?: string | null;
    title: string;
    description?: string | null;
    status: string;
    url?: string | null;
    assignee?: ExternalIssueActorModel | null;
    reporter?: ExternalIssueActorModel | null;
    sourceColumn?: string | null;
    boardColumnId?: string | null;
    createdAt?: string | null;
    updatedAt?: string | null;
    labels: string[];
};

export type ExternalIssueCommentModel = {
    id: string;
    kind: string;
    author?: ExternalIssueActorModel | null;
    body: string;
    createdAt?: string | null;
    updatedAt?: string | null;
};

export type ExternalIssueActivityEntryModel = {
    id: string;
    type: string;
    description: string;
    createdAt?: string | null;
    actor?: ExternalIssueActorModel | null;
};

export type ExternalIssueDetailsModel = {
    issue: ExternalIssueModel;
    comments: ExternalIssueCommentModel[];
    activity: ExternalIssueActivityEntryModel[];
};

File: src/Meshboard.Core/Issues/ExternalIssue.cs
namespace Meshboard.Core.Issues
{
    public class ExternalIssue
    {
        public required string ExternalId { get; set; }

        public required string IssueNumber { get; set; }

        public required string SourceKey { get; set; }

        public required string Title { get; set; }

        public string? Description { get; set; }

        public string Status { get; set; } = "open";

        public string? Url { get; set; }

        public ExternalIssueActor? Assignee { get; set; }

        public ExternalIssueActor? Reporter { get; set; }

        public string? SourceColumn { get; set; }

        public string? BoardColumnId { get; set; }

        public DateTimeOffset? UpdatedAt { get; set; }

        public DateTimeOffset? CreatedAt { get; set; }

        public IReadOnlyList<string> Labels { get; set; } = [];
    }
}